azure: support standalone (non VMSS) instances

bpineau opened the pull request:
Support instances that are not part of a Virtual Machine Scale Set.

Cilium only assumption with regard to machines running in VMSS was
limited to interfaces discovery. Also listing standard (non-VMSS)
network interfaces from the same Azure ResourceGroup is all we need
for those to work with Cilium.

```release-note
Azure: support non VMSS instances
```

bpineau committed:
azure: support standalone (non VMSS) instances

Support instances that are not part of a Virtual Machine Scale Set.

Cilium only assumption with regard to machines running in VMSS was
limited to interfaces discovery. Also listing standard (non-VMSS)
network interfaces from the same Azure ResourceGroup is all we need
for those to work proper with Cilium.

Signed-off-by: Benjamin Pineau <benjamin.pineau@datadoghq.com>

ungureanuvladvictor commented on pkg/azure/api/api.go:
```diff
@@ -101,8 +101,52 @@ func deriveStatus(err error) string {
-// describeNetworkInterfaces lists all Azure Interfaces
+// describeNetworkInterfaces lists all Azure Interfaces in the client's resource group
 func (c *Client) describeNetworkInterfaces(ctx context.Context) ([]network.Interface, error) {
+	networkInterfaces, err := c.describeVMSSNetworkInterfaces(ctx)
```
Mind renaming this `vmssNetworkInterfaces` ?

bpineau committed:
azure: provide IP allocations on standalone instances

Standalone (non VMSS) instances are slightly different than VMSS in
that regard: the interfaces are attached to instances, not part of
the instance definition.

Signed-off-by: Benjamin Pineau <benjamin.pineau@datadoghq.com>

ungureanuvladvictor commented on pkg/azure/api/api.go:
```diff
@@ -358,3 +410,38 @@ func (c *Client) AssignPrivateIpAddresses(ctx context.Context, instanceID, vmssN
+		return fmt.Errorf("unable to update interface %s: %s", interfaceName, err)
+	}
+
+	if err := future.WaitForCompletionRef(context.TODO(), c.interfaces.Client); err != nil {
```
Use the provided `ctx` that is passed to `assignPrivateIpAddressesToVM`

ungureanuvladvictor commented on pkg/azure/api/api.go:
```diff
@@ -358,3 +410,38 @@ func (c *Client) AssignPrivateIpAddresses(ctx context.Context, instanceID, vmssN
 }
+
+// assignPrivateIpAddressesToVM assign a private IP to an interface attached to a standalone instance
+func (c *Client) assignPrivateIpAddressesToVM(ctx context.Context, instanceID, vmssName, subnetID, interfaceName string, addresses int) error {
```
`vmss` param is unused, please remove it.

ungureanuvladvictor replied:
`instanceID` is unused as well, please remove it.

ungureanuvladvictor commented on pkg/azure/types/types.go:
```diff
@@ -145,6 +146,13 @@ func (a *AzureInterface) extractIDs() {
 		}
+	// Interface from a standalone instance:
+	// //subscriptions/xxx/resourceGroups/yyy/providers/Microsoft.Network/networkInterfaces/iii
+	case strings.Contains(a.ID, "/Microsoft.Network/"):
```
Mind adding a couple of test cases for this via https://github.com/cilium/cilium/blob/master/pkg/azure/types/types_test.go#L75?

bpineau replied:
Done

ungureanuvladvictor commented on pkg/azure/api/api.go:
```diff
@@ -310,6 +354,14 @@ func generateIpConfigName() string {
 // AssignPrivateIpAddresses assigns the IPs to the interface as specified by
 // the interfaceName.
 func (c *Client) AssignPrivateIpAddresses(ctx context.Context, instanceID, vmssName, subnetID, interfaceName string, addresses int) error {
+	if vmssName == "" {
```
I'm not sure this is the right place to fork the code and do the `vm` vs. `vmss` logic. `AssignPrivateIpAddresses` gets called via https://github.com/cilium/cilium/blob/master/pkg/azure/ipam/node.go#L141. 

How about we change the internal Azure API shim to be:
- `AssignPrivateIpAddressesVM`
- `AssignPrivateIpAddressesVMSS` -- this one following the implementation on the master branch currently

`AssignPrivateIpAddressesVM` -- to be called in https://github.com/cilium/cilium/blob/master/pkg/azure/ipam/node.go#L135 based on the `vmss` being not `""` in the `iface` object we cast via https://github.com/cilium/cilium/blob/master/pkg/azure/ipam/node.go#L136.

The current interface for the internal Azure shim is at https://github.com/cilium/cilium/blob/master/pkg/azure/ipam/instances.go#L29-L34.

@tgraf wdyt about this proposal?

I think another overhaul we could do is to piggy back just on the `vmID` field from https://github.com/cilium/cilium/blob/master/pkg/azure/types/types.go#L122-L123 if we set this correctly as Azure resource names. I believe that one to have the `vmss` name in it and we just have to write some extraction functions. But this is another story which we could tackle later if the code gets too confusing.

bpineau replied:
Yes I also considered that option, and don't have much preference either way.

Chose to keep `AssignPrivateIpAddresses` to remain consistent with / comparable to [`EC2API` interface](https://github.com/cilium/cilium/blob/master/pkg/aws/eni/instances.go#L42) (eg. hints what remains to be implemented for a given cloud provider, and perhaps might be later a base for a common cloud API interface).

Not sure if keeping AWS and Azure APIs similar really matters though?
Let me know what's the preferred way.

ungureanuvladvictor replied:
I think they are different and we never had the discussion to consolidate into one shim. My preference is to not do that given all the subtle differences between AWS/Azure (and also GCP for the future).

bpineau replied:
@ungureanuvladvictor implemented your suggestion

ungureanuvladvictor reviewed (changes requested):
@bpineau did you run this in a real Azure environment? Currently we do not have e2e tests for the Azure IPAM.

ungureanuvladvictor reviewed (commented):
`ctx` change + unit tests look good. Let's keep the discussion a bit open on the Azure API change.

@bpineau for `first-interface-index` but how will you know which index the interface is attached at in Azure? I think in Azure the only thing that identifies the interface on the machine is the `name` of the interface.
